# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsContentUtils.cpp
# Commit: f17be2edac14
# Full Hash: f17be2edac145591c3559810d5023f83ca0d28d0
# Author: Kris Maglione <maglione.k@gmail.com>
# Date: 2020-07-01 03:27:11
# Regressor Bug: 1646513
# File Overlap Count: 1
# Description:
#   Bug 1646513: Fix GetInProcessParentDocument usage in CheckForSubFrameDrop. r=nika
#   
#   We only allow drops from descendant frames into ancestors if they're same
#   origin anyway, which implies same-process, but outside of Fission, we allow
#   cross-origin interstitial frames, whereas with Fission we currently don't due
# ==============================================================================

diff -r a3cc583cc2ba -r f17be2edac14 dom/base/nsContentUtils.cpp
--- a/dom/base/nsContentUtils.cpp	Tue Jun 30 20:14:08 2020 +0000
+++ b/dom/base/nsContentUtils.cpp	Thu Jun 25 21:01:51 2020 +0000
@@ -5555,14 +5555,9 @@
     return true;
   }
 
-  Document* targetDoc = target->OwnerDoc();
-  nsPIDOMWindowOuter* targetWin = targetDoc->GetWindow();
-  if (!targetWin) {
-    return true;
-  }
-
   // Always allow dropping onto chrome shells.
-  if (targetWin->GetBrowsingContext()->IsChrome()) {
+  BrowsingContext* targetBC = target->OwnerDoc()->GetBrowsingContext();
+  if (targetBC->IsChrome()) {
     return false;
   }
 
@@ -5573,13 +5568,13 @@
   if (doc) {
     // Get each successive parent of the source document and compare it to
     // the drop document. If they match, then this is a drag from a child frame.
-    do {
-      doc = doc->GetInProcessParentDocument();
-      if (doc == targetDoc) {
-        // The drag is from a child frame.
+    for (BrowsingContext* bc = doc->GetBrowsingContext()->GetParent(); bc;
+         bc = bc->GetParent()) {
+      if (bc == targetBC) {
+        // The drag is from a descendant frame.
         return true;
       }
-    } while (doc);
+    }
   }
 
   return false;
